File: examples/part3-confgen-go/confgen.go
// Package main is the Go implementation of the Part 3 config-generator
// project. All three languages emit byte-identical output for the same spec.
//
//	go test ./...
//	go build && ./confgen service.conf
package main

import (
	"errors"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// Spec is a validated service spec.
type Spec struct {
	Name     string
	Domain   string
	Port     int
	Replicas int
}

// ErrMissing and ErrInvalid wrap precise spec-file problems.
var (
	ErrMissing = errors.New("missing required key")
	ErrInvalid = errors.New("invalid value")
)

// ParseSpec parses the `key = value` format (# comments, blank lines ok).
func ParseSpec(text string) (Spec, error) {
	spec := Spec{Replicas: 1} // default
	var haveName, haveDomain, havePort bool

	for _, raw := range strings.Split(text, "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, value, found := strings.Cut(line, "=")
		if !found {
			continue // tolerated: not a key=value line
		}
		key, value = strings.TrimSpace(key), strings.TrimSpace(value)
		switch key {
		case "name":
			spec.Name, haveName = value, true
		case "domain":
			spec.Domain, haveDomain = value, true
		case "port":
			p, err := strconv.Atoi(value)
			if err != nil {
				return Spec{}, fmt.Errorf("port=%q: %w", value, ErrInvalid)
			}
			spec.Port, havePort = p, true
		case "replicas":
			r, err := strconv.Atoi(value)
			if err != nil {
				return Spec{}, fmt.Errorf("replicas=%q: %w", value, ErrInvalid)
			}
			spec.Replicas = r
		}
	}

	switch {
	case !haveName:
		return Spec{}, fmt.Errorf("name: %w", ErrMissing)
	case !haveDomain:
		return Spec{}, fmt.Errorf("domain: %w", ErrMissing)
	case !havePort:
		return Spec{}, fmt.Errorf("port: %w", ErrMissing)
	case spec.Replicas < 1:
		return Spec{}, fmt.Errorf("replicas=%d: %w", spec.Replicas, ErrInvalid)
	}
	return spec, nil
}

// RenderNginx renders the upstream + server block.
func RenderNginx(spec Spec) string {
	var servers strings.Builder
	for i := 0; i < spec.Replicas; i++ {
		fmt.Fprintf(&servers, "    server 127.0.0.1:%d;\n", spec.Port+i)
	}
	return fmt.Sprintf(`upstream %[1]s {
%[3]s}

server {
    listen 80;
    server_name %[2]s;

    location / {
        proxy_pass http://%[1]s;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
    }
}
`, spec.Name, spec.Domain, servers.String())
}

// RenderSystemd renders the unit template (%i = instance).
func RenderSystemd(spec Spec) string {
	return fmt.Sprintf(`[Unit]
Description=%[1]s service (instance %%i)
After=network.target

[Service]
ExecStart=/usr/local/bin/%[1]s --port %%i
Restart=on-failure
User=%[1]s
NoNewPrivileges=true

[Install]
WantedBy=multi-user.target
`, spec.Name)
}

// Generate produces the full CLI output shared across the languages.
func Generate(text string) (string, error) {
	spec, err := ParseSpec(text)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("--- nginx: %s.conf\n%s\n--- systemd: %s@.service\n%s",
		spec.Name, RenderNginx(spec), spec.Name, RenderSystemd(spec)), nil
}

func main() {
	var data []byte
	var err error
	if len(os.Args) > 1 {
		data, err = os.ReadFile(os.Args[1])
	} else {
		data, err = io.ReadAll(os.Stdin)
	}
	if err == nil {
		var out string
		out, err = Generate(string(data))
		if err == nil {
			fmt.Print(out)
			return
		}
	}
	fmt.Fprintln(os.Stderr, "confgen:", err)
	os.Exit(1)
}

File: examples/part3-confgen-py/confgen.py
"""Config generator — Python implementation of the Part 3 project. All three
languages emit byte-identical output for the same spec.

Run the tests / the tool:

    python3 -m unittest discover -s . -p 'test_*.py'
    python3 confgen.py service.conf
"""

from __future__ import annotations

import sys
from dataclasses import dataclass


class SpecError(Exception):
    """A precise spec-file problem (missing key or invalid value)."""


@dataclass(frozen=True)
class Spec:
    """A validated service spec."""

    name: str
    domain: str
    port: int
    replicas: int


def parse_spec(text: str) -> Spec:
    """Parses the `key = value` format (# comments, blank lines ok)."""
    values: dict[str, str] = {}
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        if "=" not in line:
            continue  # tolerated: not a key=value line
        key, _, value = line.partition("=")
        values[key.strip()] = value.strip()

    for required in ("name", "domain", "port"):
        if required not in values:
            raise SpecError(f"missing required key {required}")

    def as_int(key: str, default: int | None = None) -> int:
        raw_value = values.get(key)
        if raw_value is None:
            assert default is not None
            return default
        try:
            return int(raw_value)
        except ValueError:
            raise SpecError(f"invalid value {raw_value!r} for key {key}") from None

    spec = Spec(
        name=values["name"],
        domain=values["domain"],
        port=as_int("port"),
        replicas=as_int("replicas", default=1),
    )
    if spec.replicas < 1:
        raise SpecError("invalid value '0' for key replicas")
    return spec


def render_nginx(spec: Spec) -> str:
    """Renders the upstream + server block."""
    servers = "".join(
        f"    server 127.0.0.1:{spec.port + i};\n" for i in range(spec.replicas)
    )
    return f"""upstream {spec.name} {{
{servers}}}

server {{
    listen 80;
    server_name {spec.domain};

    location / {{
        proxy_pass http://{spec.name};
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
    }}
}}
"""


def render_systemd(spec: Spec) -> str:
    """Renders the unit template (%i = instance)."""
    return f"""[Unit]
Description={spec.name} service (instance %i)
After=network.target

[Service]
ExecStart=/usr/local/bin/{spec.name} --port %i
Restart=on-failure
User={spec.name}
NoNewPrivileges=true

[Install]
WantedBy=multi-user.target
"""


def generate(text: str) -> str:
    """Full CLI output: both artifacts with headers (shared across languages)."""
    spec = parse_spec(text)
    return (
        f"--- nginx: {spec.name}.conf\n{render_nginx(spec)}\n"
        f"--- systemd: {spec.name}@.service\n{render_systemd(spec)}"
    )


def main() -> None:
    if len(sys.argv) > 1:
        with open(sys.argv[1], encoding="utf-8") as f:
            text = f.read()
    else:
        text = sys.stdin.read()
    try:
        print(generate(text), end="")
    except SpecError as e:
        print(f"confgen: {e}", file=sys.stderr)
        raise SystemExit(1) from None


if __name__ == "__main__":
    main()
